lib/zone.ts

```diff
@@ -1858,5 +1858,154 @@ const Zone: ZoneType = (function(global: any) {
+      return parentZoneDelegate.scheduleTask(targetZone, task);
+    },
+    onInvokeTask: (parentZoneDelegate, currentZone, targetZone, task, applyThis, applyArgs) => {
+      handleDetectError(Error('onInvokeTask'));
+      return parentZoneDelegate.invokeTask(targetZone, task, applyThis, applyArgs);
+    },
+    onCancelTask: (parentZoneDelegate, currentZone, targetZone, task) => {
+      handleDetectError(Error('onCancelTask'));
+      return parentZoneDelegate.cancelTask(targetZone, task);
+    },
+
+    onHasTask: (delegate, current, target, hasTaskState) => {
+      handleDetectError(Error('onHasTask'));
+      return delegate.hasTask(target, hasTaskState);
+    },
+
+    onHandleError(parentDelegate, currentZone, targetZone, error) {
+      parentDelegate.handleError(targetZone, error);
+      handleDetectError(error);
+      return false;
+    }
+  });
+
+  let detectFn = () => {
+    throw new Error('zoneAwareFrames');
+  };
+
+  let detectPromiseFn = () => {
+    new Promise((resolve, reject) => {
+      reject(new Error('zoneAwareFrames'));
+    });
+  };
+
+  let detectPromiseCaughtFn = () => {
+    const p = new Promise((resolve, reject) => {
+      reject(new Error('zoneAwareFrames'));
+    });
+    p.catch(err => {
+      throw err;
+    });
+  };
+
+  // Cause the error to extract the stack frames.
+  detectEmptyZone.runTask(
+      detectEmptyZone.scheduleEventTask('detect', detectFn, null, () => null, null));
+  detectZoneWithCallbacks.runTask(
+      detectZoneWithCallbacks.scheduleEventTask('detect', detectFn, null, () => null, null));
+  detectEmptyZone.runTask(
+      detectEmptyZone.scheduleMacroTask('detect', detectFn, null, () => null, null));
+  detectZoneWithCallbacks.runTask(
+      detectZoneWithCallbacks.scheduleMacroTask('detect', detectFn, null, () => null, null));
+  detectEmptyZone.runTask(detectEmptyZone.scheduleMicroTask('detect', detectFn, null, () => null));
+  detectZoneWithCallbacks.runTask(
+      detectZoneWithCallbacks.scheduleMicroTask('detect', detectFn, null, () => null));
+
+  detectEmptyZone.runGuarded(() => {
+    detectEmptyZone.run(detectFn);
+  });
+  detectZoneWithCallbacks.runGuarded(() => {
+    detectEmptyZone.run(detectFn);
+  });
+
+  detectEmptyZone.runGuarded(detectPromiseFn);
+  detectZoneWithCallbacks.runGuarded(detectPromiseFn);
+
+  detectEmptyZone.runGuarded(detectPromiseCaughtFn);
+  detectZoneWithCallbacks.runGuarded(detectPromiseCaughtFn);
+
+  // some functions are not easily to be detected here,
+  // for example Timeout.ZoneTask.invoke, if we want to detect those functions
+  // by detect zone, we have to run all patched APIs, it is too risky
+  // so for those functions, just check whether the stack contains the string or not.
+  const otherZoneAwareFunctionNames = [
+    'ZoneTask.invoke', 'ZoneAware', 'getStacktraceWithUncaughtError', 'new LongStackTrace',
+    'long-stack-trace'
+  ];
+
+  Object.defineProperty(ZoneAwareError, 'getNonZoneAwareStack', {
```

Comment:
1) Why is this called `getNonZoneAwareStack`?  what is returned is ZoneAware. 
Why do we even need this method? 


2) Why is this `Object.defineProperty` rather than simple function assignment?